# ProxyTyper: Proxy-panel Generation for Outsourcing Genotype Imputation 

This repository contains the source code, documentation, and an example for generating and using proxy panels for genotype imputation.

ProxyTyper uses numerous mechanisms to protect genotype data.

## Installation
To build ProxyTyper, run *make* under main directory:
```
make clean
make
```

This should build the executable under *bin/* directory.

## Genotype Imputation Pipeline

Genotype imputation tools such as *BEAGLE* and *minimac4* use sparsely genotyped (or typed in short) variants (e.g., genotyping arrays) and impute variants that were untyped in the query panel.

To achieve this, tools use a reference panel that contains a large number of variants, overlap the typed variants in the client's panel, and impute the remaining variants that are untyped.

Most popular imputation tools are based on a hidden Markov model (HMM) that is described by Li-Stephens model.

### Re-Sampling Panels
Resampling uses a stochastic resampling procedure to generate a new panel.

#### Resampling typed variant genotypes (Query Site)
Typed variant resampling generates a new proxy panel that is generated by Markov chain resampling of the typed variant haplotypes.

```
op_prefix=Query_Site
resampled_size=2000

# This is the relative effective population size while resampling. Higher value increases the probability of recombinations.
N_e_frac=0.125

# Error introduction is not used.
allele_eps=0

# Maximum segment length in base pairs, centiMorgan, # variants.
max_l_seg_n_bps=10000000
max_l_seg_cM=0
max_l_seg_nvars=0

# Number of threads.
n_threads=40

# Start and end positions for sampling.
start_posn=0
end_posn=250000000

./ProxyTyper.sh -resample_query_genotypes_w_length_cutoff ${SITE1_haplocoded_tag_matbed} ${SITE1_sample_list} \
	${resampled_size} ${per_chrom_maps_dir} ${N_e_frac} ${allele_eps} \
	${max_l_seg_n_bps} ${max_l_seg_cM} ${max_l_seg_nvars} \
	${n_threads} ${op_prefix}
```
Note that this takes only typed (tag) variants as 

#### Resampling typed and untyped variant genotypes (Reference Site)
Resampling the typed and untyped variants together uses the corresponding option:

```
op_prefix=${DATA_DIR}/SITE2_${chr_id}
resampled_size=2000
N_e_frac=0.125
allele_eps=0
max_l_seg_n_bps=10000000
max_l_seg_cM=0
max_l_seg_nvars=0
n_threads=10
start_posn=0
end_posn=250000000

./ProxyTyper.sh -resample_tag_target_genotypes_w_length_cutoff ${SITE2_haplocoded_tag_matbed} ${SITE2_haplocoded_target_matbed} ${SITE2_sample_list} \
	${resampled_size} ${per_chrom_maps_dir} ${N_e_frac} ${allele_eps} \
	${max_l_seg_n_bps} ${max_l_seg_cM} ${max_l_seg_nvars} \
	${n_threads} ${op_prefix} 

cp ${op_prefix}_resampled_tags.matbed.gz ${chr_id}_SITE2_resampled_tag_haplocoded.matbed.gz
cp ${op_prefix}_resampled_targets.matbed.gz ${chr_id}_SITE2_resampled_target_haplocoded.matbed.gz

cp ${op_prefix}_resampled_sample_ids.list ${chr_id}_SITE2_resampled_samples.list

```


### Protecting Typed Variant Alleles

- Generate the models for hashing the typed variants.

```
# Generate the permutation proxization mapping regions.
ProxyTyper -generate_save_per_site_mixing_parameters_LD_aware ${SITE1_haplocoded_tag_matbed} ${SITE1_sample_list} ${filter_proxization_vicinity_size} ${var_weight_prob} ${var2var_interaction_prob} ${var2var2var_interaction_prob} ${filter_weight_inversion_prob} ${coding_modulus} ${normalized_N_e} ${filter_proxization_min_n_params_per_var} ${per_chrom_maps_dir} ${weight_params_file}

# Generate the permutation proxization mapping regions.
ProxyTyper -generate_permute_proxizing_parameters ${SITE1_haplocoded_tag_matbed} ${SITE1_sample_list} ${perm_proxization_vicinity_size} ${perm_proxy_geno_inversion_prob} ${permute_proxying_mapping_BED}
```

- Hash the typed variants.

```
ProxyTyper -MT_proxize_variants_per_vicinity_non_linear_modular_average_per_var_custom_filters site1_resampled_panel_perm_prox.matbed.gz site1_resampled_panel_sample_ids.list ${weight_params_file} ${allele_err_eps} ${N_THREADS} site1_proxized_resampled_panel_tag_haplocoded.matbed.gz
```

### Protecting Un-Typed Variant Alleles (Reference Site)

- Permute the untyped variants; this option permutes the target variants within typed-typed variant blocks:
```
# Proxize the target variants by permutation: Note that this is done only at SITE2.
untyped_recoding_op_prefix="site2_resampled_panel_target_anon_coord"
untyped_proxy_2_target_mapping_BED=${untyped_recoding_op_prefix}_target_proxy_mapping.bed
untyped_proxized_haplocoded_target_matbed=${untyped_recoding_op_prefix}_proxized_targets.matbed.gz

untyped_var_perm_n_vicinity=20
untyped_var_allele_switch_prob=0.5

# Recode the targets for the site2: Use the anon-coord for this since we are operating on anonymized coordinates.
ProxyTyper -recode_untyped_variant_reference_panel_per_target_permutation \
	site2_proxized_resampled_panel_tag_anon_coord_haplocoded.matbed.gz site2_resampled_target_anon_coord_haplocoded.matbed.gz \
	site2_resampled_panel_sample_ids.list \
	${untyped_var_perm_n_vicinity} \
	${untyped_var_allele_switch_prob} \
	${untyped_recoding_op_prefix}

# Check the files.
if [[ ! -f ${untyped_proxy_2_target_mapping_BED} ]]
then
	echo "Could not find encoding coordinates @ \"${untyped_proxy_2_target_mapping_BED}\""
	exit 1
fi

if [[ ! -f ${untyped_proxized_haplocoded_target_matbed} ]]
then
	echo "Could not find encoded target genotypes @ \"${untyped_proxized_haplocoded_target_matbed}\""
	exit 1
fi
```

- Partition and shuffle the untyped variants:
```
# Shuffling is always hardcoded to increase privacy.
shuffle_decomp_vars=1

site2_untyped_decomp_prefix="site2_resampled_panel_target_anon_coord"

# Decompose the current untyped variants, this does not change the typed variants, which are proxied above.
ProxyTyper -simple_decompose_untyped_variants site2_proxized_resampled_panel_tag_anon_coord_haplocoded.matbed.gz \
	${untyped_proxized_haplocoded_target_matbed} \
	site2_resampled_panel_sample_ids.list \
	${variant_decomp_min_AAF} \
	${shuffle_decomp_vars} ${site2_untyped_decomp_prefix}
```

- Check the partitioning mapping region names:
```
untyped_decomposing_interval=${site2_untyped_decomp_prefix}_untyped_decomposing.interval
untyped_decomposed_target_matbed=${site2_untyped_decomp_prefix}_decomposed.matbed.gz

if [[ ! -f ${untyped_decomposing_interval} ]]
then
	echo "Could not find the decomposing interval file @ ${untyped_decomposing_interval}"
	exit 1
fi

if [[ ! -f ${untyped_decomposed_target_matbed} ]]
then
	echo "Could not find the decompose untyped genotypes file @ ${untyped_decomposed_target_matbed}"
	exit 1
fi
```

### Protecting Variant Positions and Genetic Maps (Query and Reference Sites)

This option generates uniformly spaced variants, and censors and locally anonymizes the genetic maps.

- Extract the typed/untyped variant coordinates:
```
anon_genetic_maps_dir=anon_coords_data
beagle_genetic_map_file=${anon_genetic_maps_dir}/anon_beagle_map.map

# Dump the original tag and target variants.
variation_tools -dump_plain_geno_signal_regions ${SITE1_haplocoded_tag_matbed} ${SITE1_sample_list} 1 tag_vars.bed
variation_tools -dump_plain_geno_signal_regions ${SITE1_haplocoded_target_matbed} ${SITE1_sample_list} 1 target_vars.bed
```

- Write the uniformized mapping coordinates; censor genetic map to include only the typed variants and anonymize the genetic map:
```
rm -f -r ${anon_genetic_maps_dir}
mkdir ${anon_genetic_maps_dir}
ProxyTyper -anonymize_tag_target_genetic_map_coords tag_vars.bed target_vars.bed ${per_chrom_maps_dir}/${chr_id}.map ${coord_anon_cM_noise_SD} tag_target_mapping ${anon_genetic_maps_dir}
```

- Format genetic map file to be used for imputation:
```
# Write the beagle formatted genetic map file.
awk -v chr_id=${chr_id} {'if(NR>1){print chr_id"\t.\t"$3"\t"$1}'} ${anon_genetic_maps_dir}/${chr_id}.map > ${beagle_genetic_map_file} 
```

- Use generic mapping option to map the typed variants to uniform coordinates:
```
# Map the coordinates for both site1 and site2.
ProxyTyper -generic_map_genotype_regs_per_src_dest_regs site1_proxized_resampled_panel_tag_haplocoded.matbed.gz site1_resampled_panel_sample_ids.list ${anon_genetic_maps_dir}/tag_target_mapping_original_coords.bed ${anon_genetic_maps_dir}/tag_target_mapping_mapping_coords.bed site1_proxized_resampled_panel_tag_anon_coord_haplocoded.matbed.gz

ProxyTyper -generic_map_genotype_regs_per_src_dest_regs site2_proxized_resampled_panel_tag_haplocoded.matbed.gz site2_resampled_panel_sample_ids.list ${anon_genetic_maps_dir}/tag_target_mapping_original_coords.bed ${anon_genetic_maps_dir}/tag_target_mapping_mapping_coords.bed site2_proxized_resampled_panel_tag_anon_coord_haplocoded.matbed.gz
```

- Map the untyped variant coordinates.
```
# Anonymize coordinates of target variants. Note that these are further permuted below.
ProxyTyper -generic_map_genotype_regs_per_src_dest_regs ${SITE2_haplocoded_resampled_panel_target_matbed} site2_resampled_panel_sample_ids.list ${anon_genetic_maps_dir}/tag_target_mapping_original_coords.bed ${anon_genetic_maps_dir}/tag_target_mapping_mapping_coords.bed site2_resampled_target_anon_coord_haplocoded.matbed.gz
```

---

Please refer to the *example/* directory that runs the whole imputation protocol with the steps described in the manuscript.